bin/oneoff/import_unit_details.rb

```diff
@@ -82,22 +91,104 @@ def main(options)
 
     cb_unit = JSON.parse(cb_unit_json)
     validate_unit(script, cb_unit)
+    chapters = get_validated_chapters(cb_unit)
+    lesson_pairs = get_validated_lesson_pairs(script, chapters)
+
+    log "validated unit #{script.name} data"
+
+    next if options.dry_run
+
+    lesson_pairs.each do |lesson, cb_lesson|
+      lesson.update_from_curriculum_builder(cb_lesson)
+    end
+    log "updated #{lesson_pairs.count} lessons in unit #{script.name}"
   end
 end
 
 def fetch(url)
-  log "fetching unit json from #{url}"
   uri = URI(url)
   response = Net::HTTP.get_response(uri)
   raise "HTTP status #{response.code} fetching #{uri}" unless response.is_a? Net::HTTPSuccess
   body = response.body
-  log "received #{body.length} bytes of unit json."
+  log "fetched #{body.length} bytes of unit json from #{uri}"
   body
 end
 
 def validate_unit(script, cb_unit)
   raise "unexpected unit_name #{cb_unit['unit_name']}" unless cb_unit['unit_name'] == script.name
-  log "validated unit data for #{script.name}"
+end
+
+def get_validated_chapters(cb_unit)
+  # In 2020, CSF and CSD lessons are all inside chapters, and CSP does not use
+  # chapters. Therefore, we can simplify the merge logic by assuming that
+  # lessons are all inside chapters when chapters are present.
+  if cb_unit['chapters'].present? && cb_unit['lessons'].present?
+    raise "found #{cb_unit['lessons'].count} unexpected lessons outside of chapters"
+  end
+
+  if cb_unit['chapters'].blank? && cb_unit['lessons'].blank?
+    raise "no chapters or lessons found"
+  end
+
+  cb_unit['chapters'].presence || [{'lessons' => cb_unit['lessons']}]
+end
+
+def get_validated_lesson_pairs(script, chapters)
+  validated_lesson_pairs = []
+
+  cb_lessons = chapters.map {|ch| ch['lessons']}.flatten
+
+  # Compare non-lockable lessons from CB and Code Studio.
+  lessons_nonlockable = script.lessons.reject(&:lockable)
+  # In 2020, a code_studio_url indicates a lockable lesson in CSP.
+  cb_lessons_nonlockable = cb_lessons.reject {|lesson| lesson['code_studio_url'].present?}
```

Comment:
CSD has lockable lessons that are not attached to lesson plans

Reply:
And CSP has some lockable lessons without lesson plans

Reply:
Right -- a code_studio_url on a CB lesson implies a lockable lesson in CSP, but a lockable lesson in CSP or CSD does not imply a corresponding CB lesson with a code_studio_url. Are we saying the same thing? If so, could my comment be improved?

Reply:
Oh nevermind my brain wasn't fully on here. For some reason I thought on line 145 you were comparing counts of lockable lessons not non_lockable. All good

Reply:
Added some comments to explain a bit better in https://github.com/code-dot-org/code-dot-org/pull/37772/commits/1c167d9ffc0adaa47c4f037cf195fec4b01dee79